### ISSUE TITLE
Core API

### ISSUE THREAD

**Issue by bnjmnt4n** [underdash/underdash#4]
Just briefly glancing through the “Core” and “More” functions, I think `difference` and `intersection` should be in the core, as they are used very commonly.

**Comment by chicoxyzzy** [comment]
totally agree

**Comment by jashkenas** [comment]
Are they really? Enlighten us. Got some examples of places you use `difference` and `intersection` in real-world code?

**Comment by jdalton** [comment]
Yep, I those set methods are pretty common (I use them in my own code regularly).

**Comment by megawac** [comment]
My thoughts on the current state

Strongly in favour of

```diff
+indexOf
+lastIndexOf
+now
+omit
+max
+min
+drop
+take
+first
+last
+restParam
+negate
-chain
-iteratee
```

Borderline. I'm not sure if these belong

```diff
+union
+intersection
+difference
-memoize
-compose
```

**Comment by akre54** [comment]
`first` and `last` really? I think I've only used them while chaining. I almost always just do `arr[0]` or `arr[arr.length - 1]`. Not pretty, but idiomatic (and makes me wistful for other rubyisms).

**Comment by megawac** [comment]
They're convenient for cases where `arr` can be null

**Comment by akre54** [comment]
How often do you use them outside of chaining though? I'd wager you almost always want to do `arr && arr[0]` instead. Utility functions are great, using the language is better.

**Comment by jridgewell** [comment]
I actually use `first` and `last` frequently (just for 1 element, not the `n` form).

Agree with removing `memoize` and `compose`, and everything in you "strong support" except `negate`.

**Comment by jdalton** [comment]
Lodash split `first`, `last` into `take`, `drop`, `takeWhile`, `takeRightWhile`, `dropWhile`, `dropRightWhile`. So the `first`, `last`, `initial`, `rest` are the basic forms.

Since the goal is a small core we should err on the side of small, it can always expand later (if needed).

**Comment by jdalton** [comment]
> Yep, I those set methods are pretty common (I use them in my own code regularly).

While `difference`, `intersection`, & `uniq` are handy. Lodash & Underscore differ on implementations so may be easier to punt them as non-core for now.

**Comment by jdalton** [comment]
I'm kicking around a combo of methods roughly (give or take a few) like those from Underscore 1.3.3 and Lodash 0.1.0 which comes in at a low 4kb or 5kb.

**Comment by jdalton** [comment]
Experimenting with a core that can pass the Backbone unit tests.

**Comment by jdalton** [comment]
In prepping for lodash v4 I've

* Removed the following aliases:
`all`, `any`, `backflow`, `collect`, `compose`, `contains`, `detect`, `foldl`, `foldr`, `head`, `include`, `inject`, `methods`, `object`, `select`, `tail`, `unique`

* Consolidated `_.sortBy` and `_.sortByAll`

* Made `_.sortByOrder` accept `"asc"` and `"desc"`

* Removed `thisArg` (`context`) support for methods opting for folks to use `_.bind`

* Renamed `_.callback` to `_.iteratee` to align with Underscore

* Split out `_.max`, `_.min`, `_.sum`, `_.sortedIndex`, `_.sortedLastIndex`, `_.uniq` into `_.maxBy`, `_.minBy`, `_.sumBy`, `_.sortedIndexBy`, `_.sortedLastIndexBy`, and `_.uniqBy` and changed it so the iteratee is provided only the `value` param

* Removed most of the `createXYZ` methods for improved readability

Next up I'm going to make `_.extend` and `_.defaults` align with Underscore's behavior and then create the smaller core. As for the smaller core I'm currently keeping this as-is for the npm package. This means kitchen sink and modules as usual for node folks. The smaller core build will come in to play for the bower/component packages and site downloads.

**Comment by jdalton** [comment]
I've removed `isDeep` params from `_.clone` and `_.flatten`.

**Comment by jdalton** [comment]
Removed IE 6/7 testing.

**Comment by jdalton** [comment]
Removed legacy support for boolean values for `options` object params

**Comment by SimenB** [comment]
> Removed `isDeep` params from `_.clone` and `_.flatten`

Will those still be available in some form? The reason I went from underscore to lodash in the first place was deep clone. :laughing: Maybe not the main reason I migrated (`jQuery.extend` can be made deep), but it was the reason I looked into migrating at all.
(It showed up when I googled why `_.clone` didn't work like I wanted it to)

Either, I'm really excited about where you're taking this, keep up the awesome work!

**Comment by megawac** [comment]
> Will those still be available in some form?

`cloneDeep` & `flattenDeep`

**Comment by SimenB** [comment]
Ah, nice! Never was a fan of boolean params anyway. Perfect solution. Thanks for the swift response!

**Comment by jdalton** [comment]
* Made "Collection" methods treat functions as objects
* Made `_.assign`, `_.extend`, `_.defaults`, & `_.merge` coerce values to objects so this, `_.extend(someNullThing, { 'a': 1 })` will return an object.

**Comment by jdalton** [comment]
I've removed pre-es5 environment support. Those wanting older environment support can load es5-shim first.

**Comment by riyadhalnur** [comment]
i noticed `_.assign` doesn't do child object assignments very well. it ends up overwriting the parent object's keys.

**Comment by jdalton** [comment]
> i noticed `_.assign` doesn't do child object assignments very well. it ends up overwriting the parent object's keys.

Correct recursive assignment is what `_.merge` is for.

**Comment by riyadhalnur** [comment]
@jdalton ah, thanks for clearing it up

**Comment by jdalton** [comment]
Ok so here's the 47 methods I've gotten to include in the 4kb build which should still support Backbone:
```js
["before", "bind", "chain", "clone", "defaults", "defer", "difference", "escape", "extend", "extendWith", "filter", "find", "first", "flatten", "forEach", "has", "identity", "includes", "indexOf", "invoke", "isArguments", "isArray", "isEmpty", "isEqual", "isFunction", "isObject", "isRegExp", "isString", "isTypedArray", "iteratee", "keys", "keysIn", "map", "mixin", "once", "pick", "property", "restParam", "result", "size", "some", "sortBy", "tap", "thru", "uniqueId", "values", "without"]
```

**Comment by adelgado** [comment]
> Removed the following aliases:
> `all, any, backflow, collect, compose, contains, detect, eq, foldl, foldr, head, include, inject, methods, object, select, tail, unique`

@jdalton Out of curiosity: what's the rationale for this?

**Comment by jdalton** [comment]
> Out of curiosity: what's the rationale for this?

To keep the API clean and focused.

**Comment by akre54** [comment]
"Penny wise and dollar foolish", as my dad would say...

**Comment by jdalton** [comment]
> "Penny wise and dollar foolish", as my dad would say...

Sometimes aliases are great but having close to 20 of them is a bit much. This also helps devs grok the API so they don't have to map that `contains` is the same ` include` is the same as `includes`.

**Comment by akre54** [comment]
But some of those aliases read better in real code than the canonical names (`some` vs `any` is a big one). Ruby seems to be doing just fine with its aliases in the language.

There's real fat trimming to be done in Lo-Dash and Underscore. Killing 20 aliases isn't really the problem.

**Comment by jdalton** [comment]
> But some of those aliases read better in real code than their canonical names (some vs any is a big one)

I dig favoring those with JS lang precedents over Ruby since this is a JS lib and devs are more likely to be familiar with the JS names.

> There's real fat trimming to be done in Lo-Dash and Underscore. Killing 20 aliases isn't really the problem.

Fat trimming is in larger issue which is being tackled by other line items. Trimming aliases is more about ensuring a more focused, uniform, & grokable API.

**Comment by jdalton** [comment]
Other subtle issues I've seen with aliases is that devs reach for them not knowing which is the primary/preferred API and end up with dependencies which just [point to others](https://github.com/lodash/lodash/blob/3.10.0-npm/collection/any.js) adding extraneous scaffolding for things like browserfiy/webpack.

**Comment by michaelficarra** [comment]
I'm for getting rid of aliases. Each function should have a single canonical name. If you want to use an alias, use underscore like this:

```js
import {some, some as any} from "underscore";
```

**Comment by akre54** [comment]
That's not the point at all. But fine, killing aliases is the definition of bikeshedding. Don't we have some actual culling to do?

**Comment by jdalton** [comment]
> hat's not the point at all. But fine. Killing aliases is the definition of bikeshedding. Don't we have some actual culling to do?

On my end I've created a 4kb lodash core build.

**Comment by CrossEye** [comment]
[This parable][pa] was what convinced much of my team to eliminate aliases from our codebase.  On occasion, I still miss them, but on balance I'm quite glad we did so.


  [pa]: https://github.com/ramda/ramda/pull/574#issuecomment-64124850

**Comment by jdalton** [comment]
The core build [is up](https://github.com/lodash/lodash/blob/master/lodash.core.js).

**Comment by jdalton** [comment]
The core build is composed of methods required by Backbone and others I could fit in to 4kb gzipped.
I made sure to include `reduce` since it's the swiss army knife of methods and helpers like `noConflict`.

I got the file size down by removing deep path look ups, removing array and method composition optimizations, removing features from `clone` and `sortBy` to be on par with Underscore.

**Comment by CrossEye** [comment]
:100:

Very nice!

**Comment by shawnExplore** [comment]
I think some of those 'aliases' were well established long before Lodash (e.g. head, tail, etc.), no? It would be a shame for someone coming from FP land would have to learn new terms here?

**Comment by jdalton** [comment]
They are rooted in other langs and I've fine dropping them. This is JS (this is Spartaaaa) I'm fine with folks learning alternative APIs.

**Comment by bjmiller** [comment]
Two thoughts.

1) I like the "We're all adults here" approach.  It can be made real with a sort of companion to `_.mixin`.
```javascript
_.alias({reduce: 'foldl', some: 'any'});
```
(This itself can be written as a mixin, if need be.)

2) Is it realistic to provide a "lodash-migrate" library similar to jquery-migrate, for people who want to move to v4/underdash without having to break their code right away?  Having helpful console warnings like "`_.any has been removed in favor of _.some`" sure is easier than having things just fail with "`_.any is not a function`".

**Comment by jdalton** [comment]
> 2) Is it realistic to provide a "lodash-migrate" 

You mean [this...](https://www.npmjs.com/package/lodash-migrate) :)

**Comment by bjmiller** [comment]
Exactly!

Any comments on whether an `alias` method is appropriate?

**Comment by jdalton** [comment]
> Any comments on whether an alias method is appropriate?

I think use of `_.mixin` is simple enough:
```js
_.mixin({foldl: _.reduce, any: _.some});
```

**Comment by michaelficarra** [comment]
```js
_.mixin({ alias: _.mixin });
```

:sweat_smile:

**Comment by coryandrew1988** [comment]
The suggested API for `alias` is a little different from `mixin`, but I think `mixin` is better. The number of characters typed is the same, but with `mixin`, you directly reference functions on `_`, which is convenient for static analysis.

    _.alias({reduce: 'foldl', some: 'any'});

    _.mixin({reduce: _.foldl, some: _.any});

**Comment by jdalton** [comment]
Continuing to reduce overloaded methods I've split out `_.sortedUniq`, `_.sortedUniqBy`, `_.sortedIndexOf`, and `_.sortedLastIndexOf`.

**Comment by jdalton** [comment]
I've gotten the 4kb core build tested via our CI now. Since June there's been ~700 commits which is pretty active development. It's like squeezing ~3yrs of Underscore development into ~6 months.

**Comment by alundiak** [comment]
@jdalton I realize it's not yet officially published/released, right? But would it be ok, if I would have latest code in file, and I would try to replace underdash in Backbone app to replace underscore, and another project to replace lodash, so that to test if my app work :)

**Comment by d10** [comment]
@alundiak Currently we're only in discussion phase, so there's no code that you can replace Underscore/lodash with as of right now. What @jdalton has been referring to is lodash 4.0, due out in January, which will have a core build based on ideas in current discussions.

**Comment by jdalton** [comment]
Lodash v4 [release notes](https://github.com/lodash/lodash/releases/tag/4.0.0)!

**Comment by jdalton** [comment]
Closed by #14.
